File: pyBAScloudAPI/tests/reading_tests.py
import unittest
import time 
import datetime

from definitions import *

import pyBAScloudAPI as api

BCAPI = api.EntityContext(BASCLOUD_TEST_URL)


def errorHandler(e, j):
    raise e

def authenticate():
    currentDateTime = int(time.time())
    print("CURRENT DATETIME: ", currentDateTime)

    currentDateTimeString = datetime.datetime.fromtimestamp(currentDateTime).strftime("%FT%T.000Z")
    print("CURRENT DATETIME STR: ", currentDateTimeString)

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS)

    print("TOKEN: ", BCAPI.getToken())
    print("EXPIRATION DATE: ", datetime.datetime.fromtimestamp(BCAPI.getTokenExpirationDate()).strftime("%FT%T.000Z"))


class TestReadingCollection(unittest.TestCase):

    def test_reading(self):
        print("\tRequesting all readings...")

        readings = BCAPI.getReadingsCollection(BASCLOUD_TEST_TENANT_UUID, errorHandler=errorHandler)
        print("\t\tOK.")

        print("\tFound readings: ", len(readings[0]))

        self.assertTrue(len(readings[0]) >= 1)


class TestSingleReading(unittest.TestCase):

    def test_reading(self):
        print("\tRequesting single reading with UUID...")

        reading = BCAPI.getReading(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_READING_UUID)
        print("\t\tOK.")

        print("\tReading UUID: ", reading.uuid)
        print("\tReading Value: ", reading.value)
        print("\tReading timestamp: ", datetime.datetime.fromtimestamp(reading.timestamp).strftime("%FT%T.000Z"))
        print("\tReading created at: ", datetime.datetime.fromtimestamp(reading.createdAt).strftime("%FT%T.000Z"))
        print("\tReading updated at: ", datetime.datetime.fromtimestamp(reading.updatedAt).strftime("%FT%T.000Z"))


        self.assertEqual(reading.uuid, BASCLOUD_TEST_READING_UUID)
        self.assertEqual(reading.value, 1234.56)
        self.assertTrue(reading.timestamp > 0)
        self.assertTrue(reading.createdAt > 0)
        self.assertTrue(reading.updatedAt > 0)


class TestAssociatedDevice(unittest.TestCase):

    def test_reading(self):
        print("\tRequesting associated device of the reading again...")

        read_device = BCAPI.getAssociatedDevice(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_READING_UUID)
        print("\t\tOK.")

        self.assertEqual(read_device.uuid, BASCLOUD_TEST_DEVICE_UUID)



class TestCreateAndDeleteReading(unittest.TestCase):

    def test_reading(self):
        currentDateTime = int(time.time())

        new_read = BCAPI.createReading(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_DEVICE_UUID, 2345.67, currentDateTime)
        print("\t\tOK.")

        new_uuid = new_read.uuid

        self.assertEqual(new_read.value, 2345.67)
        self.assertEqual(new_read.timestamp, currentDateTime)
        # self.assertTrue(new_read.createdAt >= currentDateTime) # TODO ML: this is wrong at the moment, for some reason the backend returns timestamps -2hours
        # self.assertTrue(new_read.updatedAt() >= currentDateTime)

        
        print("\tRequesting deletion of created reading again... ")

        BCAPI.deleteReading(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        print("\t\tOK.")

        try:
            non_read = BCAPI.getReading(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        except api.NotFoundRequest:
            self.assertTrue(True)
        except:
            self.assertTrue(False)


if __name__ == '__main__':
    authenticate()
    unittest.main()

File: tests/reading_tests.cpp
#include <chrono>
#include <iostream>

#include <fmt/core.h>
#include <fmt/chrono.h>

#include <gtest/gtest.h>

#include "EntityContext.h"
#include "entity/EntityCollection.h"
#include "error/Exceptions.h"


using namespace BAScloud;

EntityContext BCAPI(BASCLOUD_TEST_URL);


void authenticate() {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    std::cerr <<"CURRENT DATETIME: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(currentDateTime))<<std::endl;

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS);

    std::cerr <<"TOKEN: "<<BCAPI.getToken()<<std::endl;
    std::cerr <<"EXPIRATION DATE: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(BCAPI.getTokenExpirationDate()))<<std::endl;
}

TEST(BasicTests, ReadingCollectionTest) {

    std::cout << "\tRequesting all readings..." << std::endl;

    EntityCollection<Reading> readings = BCAPI.getReadingsCollection(BASCLOUD_TEST_TENANT_UUID, {}, -1, -1, -1, std::numeric_limits<double>::quiet_NaN(), {}, -1, -1, [](const std::string& e, const std::string& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound readings: " << readings.first.size() << std::endl;

    EXPECT_TRUE(readings.first.size() >= 1);
}

TEST(BasicTests, ReadingCollectionPagingTest) {

    std::cout << "\tRequesting paged readings..." << std::endl;

    EntityCollection<Reading> readings = BCAPI.getReadingsCollection(BASCLOUD_TEST_TENANT_UUID, PagingOption(1000), -1, -1, -1, std::numeric_limits<double>::quiet_NaN(), BASCLOUD_TEST_DEVICE_UUID, -1, -1, [](const std::string& e, const std::string& j) {
            throw e;
        });
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound readings: " << readings.first.size() << std::endl;
    std::cout << "\tFound total pages: " << readings.second.totalPages << std::endl;

    EXPECT_TRUE(readings.first.size() >= 1);

    int page_cnt = 1;
    while(!readings.second.nextPagePointer.empty()) {
        readings = BCAPI.getReadingsCollection(BASCLOUD_TEST_TENANT_UUID, PagingOption(1000, PagingOption::Direction::NEXT, readings.second.nextPagePointer), -1, -1, -1, std::numeric_limits<double>::quiet_NaN(), BASCLOUD_TEST_DEVICE_UUID, -1, -1, [](const std::string& e, const std::string& j) {
            throw e;
        });

        std::cout << "\tFound readings: " << readings.first.size() << std::endl;
        EXPECT_TRUE(readings.first.size() >= 1);

        page_cnt++;
        
        std::cout << "\tFound paging links: Prev:" << readings.second.previousPagePointer << " Next:" << readings.second.nextPagePointer << std::endl;
    }

    EXPECT_TRUE(readings.second.totalPages == page_cnt);
}

TEST(BasicTests, SingleReadingTest) {

    std::cout << "\tRequesting single reading with UUID..." << std::endl;

    Reading reading = BCAPI.getReading(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_READING_UUID);
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tReading UUID: " << reading.getUUID() << std::endl;
    std::cout << "\tReading Value: " << reading.getValue() << std::endl;
    std::cout << "\tReading timestamp: " << fmt::format("{:%F %T}", fmt::localtime(reading.getTimestamp())) << std::endl;
    std::cout << "\tReading created at: " << fmt::format("{:%F %T}", fmt::localtime(reading.getCreatedDate())) << std::endl;
    std::cout << "\tReading updated at: " << fmt::format("{:%F %T}", fmt::localtime(reading.getLastUpdatedDate())) << std::endl;


    EXPECT_EQ(reading.getUUID(), BASCLOUD_TEST_READING_UUID);
    EXPECT_EQ(reading.getValue(), 1234.56);
    EXPECT_TRUE(reading.getTimestamp() > 0);
    EXPECT_TRUE(reading.getCreatedDate() > 0);
    EXPECT_TRUE(reading.getLastUpdatedDate() > 0);
}


TEST(BasicTests, AssociatedDeviceTest) {

    std::cout << "\tRequesting associated device of the reading again..." << std::endl;

    Device read_device = BCAPI.getAssociatedReadingsDevice(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_READING_UUID);
    std::cout << "\t\tOK." << std::endl;

    EXPECT_EQ(read_device.getUUID(), BASCLOUD_TEST_DEVICE_UUID);
}


TEST(BasicTests, CreateAndDeleteReadingTest) {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    Reading new_read = BCAPI.createReading(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_DEVICE_UUID, 2345.67, currentDateTime);
    std::cout << "\t\tOK." << std::endl;

    std::string new_uuid = new_read.getUUID();

    EXPECT_EQ(new_read.getValue(), 2345.67);
    EXPECT_EQ(new_read.getTimestamp(), currentDateTime);
    // EXPECT_TRUE(new_read.getCreatedDate() >= currentDateTime); // TODO ML: this is wrong at the moment, for some reason the backend returns timestamps -2hours
    // EXPECT_TRUE(new_read.getLastUpdatedDate() >= currentDateTime);

    
    std::cout << "\tRequesting deletion of created reading again... " << std::endl;

    BCAPI.deleteReading(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    std::cout << "\t\tOK." << std::endl;

    try {
        Reading non_read = BCAPI.getReading(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    } catch(NotFoundRequest& e) {
        EXPECT_TRUE(true);
    } catch(...) {
        EXPECT_TRUE(false);
    }

}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);

    authenticate();

    return RUN_ALL_TESTS();
}